### PR TITLE
SI-10102 REPL ignores sbt init errors

### PR DESCRIPTION
**Pull request by som-snytt**
sbt runs user init code early, while creating its
custom `IMain`. Failures in this (arbitrary) code
keep the console from starting, including (silent)
errors due to bugs. Ignore silent errors emitted
on startup when sbt is running.

### CODE REVIEW

**Comment by SethTisue** [comment]
can you suggest a reviewer who could consider this from the sbt side...?

**Comment by som-snytt** [comment]
@dwijnand still owes me one for https://github.com/scala/scala/pull/5214

**Comment by som-snytt** [comment]
Also, instead of deprecating `InterpreterLoop`, it should be considered the canonical place to workaround the sbt code. (I intended to donate upgraded `ConsoleInterface` to sbt.)

**Comment by dwijnand** [comment]
I'm not sure. On all front to be honest. On (a) what broke it, (b) what exactly happens, and (c) fixing it like that. Sorry :-/

**Comment by som-snytt** [comment]
(a) previously, I had refactored how REPL startup happens to be more sane. (b) it checks whether the reporter saw errors during startup. Since sbt executes arbitrary code during startup, that is not unlikely. (c) How about we not fail to start in that case. The reporter still knows it saw an error. Reporters like elephants never forget.

But wait, at this point, I want to refactor some imports to use trailing commas.

File: src/repl/scala/tools/nsc/interpreter/ReplReporter.scala [hunk 1]
```diff
@@ -29,8 +29,22 @@ class ReplReporter(intp: IMain) extends ConsoleReporter(intp.settings, Console.i
...
   override def warning(pos: Position, msg: String): Unit = withoutTruncating(super.warning(pos, msg))
-  override def error(pos: Position, msg: String): Unit   = withoutTruncating(super.error(pos, msg))
+  override def error(pos: Position, msg: String): Unit   = {
+    val count0 = errorCount
+    withoutTruncating(super.error(pos, msg))
+    val count1 = errorCount
+    if (count1 > count0 && intp.totalSilence)
+      silentErrors += 1
+  }
+  def reportableErrorCount = errorCount - silentErrors
```
Comment (adriaanm) [inline src/repl/scala/tools/nsc/interpreter/ReplReporter.scala]: Let's make these additions `private[interpreter]`.

File: src/repl/scala/tools/nsc/interpreter/ReplReporter.scala [hunk 2]
```diff
@@ -29,8 +29,22 @@ class ReplReporter(intp: IMain) extends ConsoleReporter(intp.settings, Console.i
...
   }
 
+  private var silentErrors = 0
   override def warning(pos: Position, msg: String): Unit = withoutTruncating(super.warning(pos, msg))
-  override def error(pos: Position, msg: String): Unit   = withoutTruncating(super.error(pos, msg))
+  override def error(pos: Position, msg: String): Unit   = {
+    val count0 = errorCount
+    withoutTruncating(super.error(pos, msg))
+    val count1 = errorCount
+    if (count1 > count0 && intp.totalSilence)
```
Comment (adriaanm) [inline src/repl/scala/tools/nsc/interpreter/ReplReporter.scala:38]: Does the reporter contract guarantee `0 <= count1 - count0 <= 1`? I guess so.

File: src/repl/scala/tools/nsc/interpreter/ReplReporter.scala (hunk 1 above)
Comment (adriaanm) [inline src/repl/scala/tools/nsc/interpreter/ReplReporter.scala]: Let's make these additions `private[interpreter]`.

**Review by adriaanm — CHANGES_REQUESTED** [review]
You've unlocked the pragmatic fix badge! Thanks for moving us in the right direction here. Eventually, closer cooperation of sbt & scalac is the ticket. A few small requested tweaks, if I may. Also, in future, please consider separate boy scout and bug fix commits.

**Comment by som-snytt** [comment]
Made `isSbt` a member overridden in `InterpreterLoop` and also pushed down `main`, which may trigger mima.

Members of `ReplReporter` are more private.

Rebased.

File: src/repl/scala/tools/nsc/InterpreterLoop.scala [hunk 3]
```diff
@@ -3,10 +3,15 @@ package scala.tools.nsc
...
-/** A compatibility stub.
+/** A compatibility stub for sbt.
  */
-@deprecated("Use a class in the scala.tools.nsc.interpreter package.", "2.9.0")
+@deprecated("Use scala.tools.nsc.interpreter.ILoop.", "2.9.0")
 class InterpreterLoop(in0: Option[BufferedReader], out: PrintWriter) extends ILoop(in0, out) {
   def this(in0: BufferedReader, out: PrintWriter) = this(Some(in0), out)
   def this() = this(None, new PrintWriter(scala.Console.out))
+
+  override protected final val isSbt = true
```
Comment (dwijnand) [inline src/repl/scala/tools/nsc/InterpreterLoop.scala:13]: Given Adriaan's previous comment, at a guess I would think this should be `protected[nsc]` (or tools or scala).

nsc is principle of least power and scala is principle of maximal flexibility?

**Review by dwijnand — APPROVED** [review]
Thanks Som for your explanation and sorry for the late review. This needs rebasing.

I've noted a little concern, if you agree then I guess you could fix it up while rebasing.

It's not a blocker for me though. LGTM.

**Comment by dwijnand** [comment]
It's a shame we can't create a test case to reproduce the original bug and stand as a regression test.

**Comment by retronym** [comment]
This PR needs a rebase to resolve a conflict with #5727

**Comment by som-snytt** [comment]
Probably I'll rebase atop `Adriaanite`?

**Comment by SethTisue** [comment]
@som-snytt rebase?

**Comment by lrytz** [comment]
Adriaanite is 2.13, so it cannot bite here. I think it would be good to get this fix in for 2.12.3, as the error feedback for affected users is non-obvious.

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala [hunk 4]
```diff
@@ -988,15 +988,15 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
           createInterpreter()
         }
         intp.initializeSynchronous()
-        globalFuture = Future successful true
-        if (intp.reporter.hasErrors) {
+        globalFuture = Future.successful(true)
+        if (intp.reporter.hasErrors && (!isSbt || intp.reporter.hasReportableErrors)) {
           echo("Interpreter encountered errors during initialization!")
```
Comment (lrytz) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:993]: Should we actually show (some of) the errors? How can I trigger this line?

**Review by lrytz — APPROVED** [review]

**Pull request merged**